File: src/FinApp.Contracts/BankFileParser.cs
using System.Globalization;
using System.Text;
using System.Xml.Linq;

namespace FinApp.Contracts;

/// <summary>One transaction parsed from an uploaded statement. <see cref="Amount"/> is signed:
/// negative = money out (an expense), positive = money in.</summary>
public readonly record struct ImportedTransaction(DateOnly Date, decimal Amount, string Description);

/// <summary>
/// Pure parsing of downloaded bank statements into <see cref="ImportedTransaction"/>s — a compliance-free,
/// zero-cost alternative to live Open Banking. Supports the two standardized formats (OFX, QIF) with no
/// per-bank mapping, plus generic CSV where the caller supplies the column indices (the UI offers a mapper).
/// No I/O and no dependencies, so it's fully unit-testable and runs the same in the WASM and MAUI hosts.
/// </summary>
public static class BankFileParser
{
    public enum Format { Unknown, Ofx, Qif, Csv, Xml, Html }

    /// <summary>Guess the format from the file name, then the content. Note some banks name an HTML-table export
    /// ".xls" and an XML export ".xml"/".camt" — content sniffing catches those.</summary>
    public static Format Detect(string? fileName, string text)
    {
        var head = TrimBom(text).TrimStart();
        // Content first for the ambiguous ones: an .xls that's really an HTML table, an .xml statement, etc.
        if (head.Contains("<OFX", StringComparison.OrdinalIgnoreCase) || head.Contains("<STMTTRN", StringComparison.OrdinalIgnoreCase)) return Format.Ofx;
        if (head.StartsWith("<html", StringComparison.OrdinalIgnoreCase) || head.Contains("<table", StringComparison.OrdinalIgnoreCase)) return Format.Html;
        if (head.Contains("BkToCstmrStmt", StringComparison.OrdinalIgnoreCase) || head.Contains(":camt.", StringComparison.OrdinalIgnoreCase)
            || head.Contains("<AccountMovement", StringComparison.OrdinalIgnoreCase)
            || (head.StartsWith('<') && head.Contains("<Ntry", StringComparison.OrdinalIgnoreCase))) return Format.Xml;

        var ext = (fileName ?? "").ToLowerInvariant();
        if (ext.EndsWith(".ofx") || ext.EndsWith(".qfx")) return Format.Ofx;
        if (ext.EndsWith(".qif")) return Format.Qif;
        if (ext.EndsWith(".htm") || ext.EndsWith(".html")) return Format.Html;
        if (ext.EndsWith(".xml") || ext.EndsWith(".camt")) return Format.Xml;
        if (ext.EndsWith(".csv") || ext.EndsWith(".tsv") || ext.EndsWith(".txt")) return Format.Csv;
        if (head.StartsWith("!Type:", StringComparison.OrdinalIgnoreCase)) return Format.Qif;
        return text.Contains(',') || text.Contains(';') || text.Contains('\t') ? Format.Csv : Format.Unknown;
    }

    private static string TrimBom(string s) => s.Length > 0 && s[0] == '﻿' ? s[1..] : s;

    // ---- ISO 20022 CAMT.053 (bank-to-customer statement) XML ----

    // Row-element names across the bank statement schemas we've seen (ISO 20022 CAMT, DAIS eBank, generic).
    private static readonly string[] XmlRowNames = { "Ntry", "AccountMovement", "Transaction", "StatementLine", "StmtLine", "Movement", "Operation" };
    private static readonly string[] XmlAmountNames = { "Amt", "Amount", "Sum", "Value" };
    private static readonly string[] XmlDirNames = { "CdtDbtInd", "MovementType", "Type", "Direction" };
    private static readonly string[] XmlDateNames = { "ValueDate", "BookgDt", "ValDt", "Date", "Dt", "DtTm" };
    private static readonly string[] XmlDescNames = { "Ustrd", "Reason", "Description", "Narrative", "AddtlTxInf", "AddtlNtryInf", "Details" };

    /// <summary>Parse a statement XML into transactions, schema-flexible so it handles ISO 20022 CAMT.053
    /// (<c>&lt;Ntry&gt;</c> + <c>CdtDbtInd</c>), the DAIS eBank format (<c>&lt;AccountMovement&gt;</c> +
    /// <c>MovementType</c>), and similar. Direction (debit/credit) sets the sign; if a schema has none, the amount's
    /// own sign is kept. Namespace-agnostic (matched by local element name).</summary>
    public static IReadOnlyList<ImportedTransaction> ParseXml(string text)
    {
        var list = new List<ImportedTransaction>();
        XDocument doc;
        try { doc = XDocument.Parse(text); } catch { return list; }

        foreach (var row in doc.Descendants().Where(e => XmlRowNames.Contains(e.Name.LocalName, StringComparer.OrdinalIgnoreCase)))
        {
            string? First(string[] names) => row.Descendants()
                .FirstOrDefault(x => names.Contains(x.Name.LocalName, StringComparer.OrdinalIgnoreCase) && !string.IsNullOrWhiteSpace(x.Value))?.Value?.Trim();

            var amtRaw = First(XmlAmountNames);
            var dateRaw = First(XmlDateNames);
            var dir = First(XmlDirNames);
            var desc = First(XmlDescNames) ?? "";
            if (amtRaw is null || dateRaw is null || !TryAmount(amtRaw, out var amount) || !TryLooseDate(dateRaw, out var date)) continue;

            // DBIT / Debit → money out; CRDT / Credit → money in. No direction element → keep the amount's own sign.
            if (dir is not null)
                amount = dir.StartsWith("C", StringComparison.OrdinalIgnoreCase) ? Math.Abs(amount) : -Math.Abs(amount);
            list.Add(new ImportedTransaction(date, amount, Clean(desc)));
        }
        return list;
    }

    // ---- HTML table (banks that export a .xls/.htm that is really an HTML <table>, e.g. DAIS eBank) ----

    /// <summary>Read an HTML <c>&lt;table&gt;</c> into a header row + data rows of plain strings — the same shape as
    /// <see cref="ReadCsv"/>, so a bank's HTML "Excel" export flows through the column mapper. Cells are de-tagged and
    /// HTML-entity-decoded; the header is the first row that has several cells (so a colspan title banner is skipped).</summary>
    public static (IReadOnlyList<string> Headers, IReadOnlyList<IReadOnlyList<string>> Rows) ReadHtmlTable(string html)
    {
        var rows = new List<IReadOnlyList<string>>();
        foreach (System.Text.RegularExpressions.Match tr in System.Text.RegularExpressions.Regex.Matches(
            html, "<tr\\b[^>]*>(.*?)</tr>", RxOptions))
        {
            var cells = new List<string>();
            foreach (System.Text.RegularExpressions.Match td in System.Text.RegularExpressions.Regex.Matches(
                tr.Groups[1].Value, "<t[dh]\\b[^>]*?/>|<t[dh]\\b[^>]*>(.*?)</t[dh]>", RxOptions))
                cells.Add(td.Groups[1].Success ? CleanHtml(td.Groups[1].Value) : "");
            if (cells.Count > 0) rows.Add(cells);
        }
        // The header is the first row with several columns (skips single-cell title/colspan banners).
        var headerIdx = rows.FindIndex(r => r.Count >= 3);
        if (headerIdx < 0) return ([], []);
        return (rows[headerIdx], rows.Skip(headerIdx + 1).Where(r => r.Count >= 3).ToList());
    }

    private const System.Text.RegularExpressions.RegexOptions RxOptions =
        System.Text.RegularExpressions.RegexOptions.Singleline | System.Text.RegularExpressions.RegexOptions.IgnoreCase;

    private static string CleanHtml(string cell)
    {
        var noTags = System.Text.RegularExpressions.Regex.Replace(cell, "<[^>]+>", " ");
        var decoded = noTags.Replace("&lt;", "<").Replace("&gt;", ">").Replace("&amp;", "&")
            .Replace("&nbsp;", " ").Replace("&quot;", "\"").Replace("&#39;", "'").Replace("&apos;", "'");
        return Clean(decoded);
    }

    // ---- OFX / QFX ----

    public static IReadOnlyList<ImportedTransaction> ParseOfx(string text)
    {
        var list = new List<ImportedTransaction>();
        foreach (var block in Between(text, "<STMTTRN>", "</STMTTRN>"))
        {
            var amtRaw = Tag(block, "TRNAMT");
            var dateRaw = Tag(block, "DTPOSTED");
            var name = Tag(block, "NAME");
            var memo = Tag(block, "MEMO");
            if (amtRaw is null || dateRaw is null) continue;
            if (!TryAmount(amtRaw, out var amount)) continue;
            if (!TryOfxDate(dateRaw, out var date)) continue;
            list.Add(new ImportedTransaction(date, amount, Clean(name ?? memo ?? "")));
        }
        return list;
    }

    private static IEnumerable<string> Between(string text, string open, string close)
    {
        var i = 0;
        while (true)
        {
            var s = text.IndexOf(open, i, StringComparison.OrdinalIgnoreCase);
            if (s < 0) yield break;
            var e = text.IndexOf(close, s, StringComparison.OrdinalIgnoreCase);
            if (e < 0) yield break;
            yield return text.Substring(s + open.Length, e - s - open.Length);
            i = e + close.Length;
        }
    }

    /// <summary>Read an OFX/SGML tag value: works whether the tag is closed (&lt;T&gt;v&lt;/T&gt;) or open
    /// (&lt;T&gt;v up to the next tag or newline), as real OFX files mix both.</summary>
    private static string? Tag(string block, string tag)
    {
        var open = "<" + tag + ">";
        var s = block.IndexOf(open, StringComparison.OrdinalIgnoreCase);
        if (s < 0) return null;
        s += open.Length;
        var end = s;
        while (end < block.Length && block[end] != '<' && block[end] != '\r' && block[end] != '\n') end++;
        return block.Substring(s, end - s).Trim();
    }

    private static bool TryOfxDate(string raw, out DateOnly date)
    {
        date = default;
        var digits = new string(raw.Where(char.IsDigit).ToArray());
        if (digits.Length < 8) return false;
        return DateOnly.TryParseExact(digits[..8], "yyyyMMdd", CultureInfo.InvariantCulture, DateTimeStyles.None, out date);
    }

    // ---- QIF ----

    public static IReadOnlyList<ImportedTransaction> ParseQif(string text)
    {
        var list = new List<ImportedTransaction>();
        DateOnly? date = null; decimal? amount = null; string payee = "", memo = "";
        foreach (var rawLine in text.Split('\n'))
        {
            var line = rawLine.TrimEnd('\r');
            if (line.Length == 0) continue;
            switch (line[0])
            {
                case 'D': if (TryLooseDate(line[1..].Trim(), out var d)) date = d; break;
                case 'T' or 'U': if (TryAmount(line[1..], out var a)) amount = a; break;
                case 'P': payee = line[1..].Trim(); break;
                case 'M': memo = line[1..].Trim(); break;
                case '^':
                    if (date is { } dd && amount is { } aa)
                        list.Add(new ImportedTransaction(dd, aa, Clean(payee.Length > 0 ? payee : memo)));
                    date = null; amount = null; payee = ""; memo = ""; break;
            }
        }
        return list;
    }

    // ---- CSV ----

    /// <summary>Split CSV/TSV text into a header row + data rows, honouring quoted fields and auto-detecting the
    /// delimiter (comma, semicolon or tab). Returns an empty header when there are no rows.</summary>
    public static (IReadOnlyList<string> Headers, IReadOnlyList<IReadOnlyList<string>> Rows) ReadCsv(string text)
    {
        var lines = text.Replace("\r\n", "\n").Split('\n').Where(l => l.Trim().Length > 0).ToList();
        if (lines.Count == 0) return ([], []);
        var delimiter = DetectDelimiter(lines[0]);
        var all = lines.Select(l => SplitCsvLine(l, delimiter)).ToList();
        return (all[0], all.Skip(1).ToList());
    }

    private static char DetectDelimiter(string header)
    {
        int c(char ch) => header.Count(x => x == ch);
        var tab = c('\t'); var semi = c(';'); var comma = c(',');
        return tab >= semi && tab >= comma && tab > 0 ? '\t' : semi > comma ? ';' : ',';
    }

    private static List<string> SplitCsvLine(string line, char delimiter)
    {
        var fields = new List<string>();
        var sb = new StringBuilder();
        var inQuotes = false;
        for (var i = 0; i < line.Length; i++)
        {
            var ch = line[i];
            if (inQuotes)
            {
                if (ch == '"' && i + 1 < line.Length && line[i + 1] == '"') { sb.Append('"'); i++; }
                else if (ch == '"') inQuotes = false;
                else sb.Append(ch);
            }
            else if (ch == '"') inQuotes = true;
            else if (ch == delimiter) { fields.Add(sb.ToString().Trim()); sb.Clear(); }
            else sb.Append(ch);
        }
        fields.Add(sb.ToString().Trim());
        return fields;
    }

    /// <summary>Turn CSV data rows into transactions using the caller-chosen column indices. Amount can come from one
    /// signed column, or from separate debit/credit columns (debit = out, credit = in). Rows that don't parse are
    /// skipped rather than throwing, so one bad line never fails a whole import.</summary>
    public static IReadOnlyList<ImportedTransaction> ParseCsv(IReadOnlyList<IReadOnlyList<string>> rows,
        int datecol, int desccol, int amountcol, int? debitcol, int? creditcol)
    {
        var list = new List<ImportedTransaction>();
        foreach (var r in rows)
        {
            string? Col(int? i) => i is { } idx && idx >= 0 && idx < r.Count ? r[idx] : null;
            if (!TryLooseDate((Col(datecol) ?? "").Trim(), out var date)) continue;

            decimal amount;
            if (debitcol is not null || creditcol is not null)
            {
                var debit = TryAmount(Col(debitcol) ?? "", out var dv) ? Math.Abs(dv) : 0m;
                var credit = TryAmount(Col(creditcol) ?? "", out var cv) ? Math.Abs(cv) : 0m;
                amount = credit - debit;
            }
            else if (!TryAmount(Col(amountcol) ?? "", out amount)) continue;
            if (amount == 0m) continue;

            list.Add(new ImportedTransaction(date, amount, Clean(Col(desccol) ?? "")));
        }
        return list;
    }

    // ---- shared helpers ----

    /// <summary>Parse a money string tolerant of currency symbols, spaces, thousands separators, both decimal
    /// conventions (1,234.56 and 1.234,56) and parenthesised negatives.</summary>
    public static bool TryAmount(string raw, out decimal amount)
    {
        amount = 0m;
        if (string.IsNullOrWhiteSpace(raw)) return false;
        var s = raw.Trim();
        var negative = s.StartsWith('(') && s.EndsWith(')') || s.Contains('-');
        var cleaned = new string(s.Where(c => char.IsDigit(c) || c is '.' or ',').ToArray());
        if (cleaned.Length == 0) return false;

        var lastDot = cleaned.LastIndexOf('.');
        var lastComma = cleaned.LastIndexOf(',');
        if (lastDot >= 0 && lastComma >= 0)
        {
            // Whichever separator is last is the decimal point; the other is a thousands separator.
            var decimalSep = lastDot > lastComma ? '.' : ',';
            var thousandsSep = decimalSep == '.' ? ',' : '.';
            cleaned = cleaned.Replace(thousandsSep.ToString(), "").Replace(decimalSep, '.');
        }
        else if (lastComma >= 0)
        {
            // Only commas: treat as decimal if it looks like one (e.g. "12,50"), else thousands.
            cleaned = cleaned.Length - lastComma - 1 <= 2 && cleaned.Count(c => c == ',') == 1
                ? cleaned.Replace(',', '.') : cleaned.Replace(",", "");
        }
        if (!decimal.TryParse(cleaned, NumberStyles.Any, CultureInfo.InvariantCulture, out amount)) return false;
        amount = Math.Abs(amount);
        if (negative) amount = -amount;
        return true;
    }

    private static readonly string[] DateFormats =
    {
        "yyyy-MM-dd", "yyyy/MM/dd", "dd/MM/yyyy", "dd.MM.yyyy", "dd-MM-yyyy",
        "MM/dd/yyyy", "d/M/yyyy", "M/d/yyyy", "yyyyMMdd", "dd/MM/yy", "MM/dd/yy",
    };

    /// <summary>Parse a date across the common bank formats. Handles a trailing time component (e.g. Revolut's
    /// "2026-07-01 12:09:37") by keeping only the date part. ISO and dd/MM are tried before MM/dd, so European
    /// statements read correctly; genuinely ambiguous values are shown to the user in the review step.</summary>
    public static bool TryLooseDate(string raw, out DateOnly date)
    {
        date = default;
        if (string.IsNullOrWhiteSpace(raw)) return false;
        var s = raw.Trim().Replace("'", "/");
        // Drop a trailing time part ("2026-07-01 12:09:37" or "2026-07-01T12:09:37") — statements often include it.
        var sep = s.IndexOfAny([' ', 'T', 't']);
        var datePart = sep > 0 ? s[..sep] : s;
        foreach (var f in DateFormats)
            if (DateOnly.TryParseExact(datePart, f, CultureInfo.InvariantCulture, DateTimeStyles.None, out date)) return true;
        if (DateOnly.TryParse(datePart, CultureInfo.InvariantCulture, DateTimeStyles.None, out date)) return true;
        // Last resort: parse the whole thing as a datetime and take the date.
        if (DateTime.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.None, out var dt)) { date = DateOnly.FromDateTime(dt); return true; }
        return false;
    }

    private static string Clean(string s) => string.Join(' ', s.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries)).Trim();
}

File: android/app/src/main/java/com/tandemtab/app/data/BankFileParser.kt
package com.tandemtab.app.data

import java.math.BigDecimal
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.Locale

/**
 * One transaction parsed from a downloaded statement. [amount] is SIGNED: negative = money out (an expense),
 * positive = money in.
 */
data class ImportedTransaction(val date: LocalDate, val amount: Double, val description: String)

/**
 * Statement parsing, on the device.
 *
 * ⚠️ **This is a port of `FinApp.Contracts.BankFileParser`, and the two have to keep agreeing.** They exist twice
 * because the parse cannot move to the server: the privacy promise is that the statement FILE never leaves the
 * phone, and only the rows the user reviewed and kept are posted to /import. Sending the file up to reuse one
 * parser would trade the promise for the convenience — so the duplication is deliberate, and the mitigation is
 * that both halves are pinned by tests over the same fixtures (see BankFileParserTest).
 *
 * ⚠️ **A parser is not a place to improvise.** Every rule below matches the C# one deliberately, including the ones
 * that look arbitrary — which separator wins when a file has both, why dd/MM is tried before MM/dd, why a bad line
 * is skipped rather than thrown on. Where behaviour differs it is stated in a comment, not left to be discovered.
 *
 * Formats: CSV/TSV (with a column mapper), OFX/QFX and QIF. XML (ISO 20022 CAMT) and the HTML-table exports some
 * banks name ".xls" are NOT ported — [detect] recognises them so the UI can say so plainly rather than presenting
 * an empty review list, which is the failure that reads as "the app is broken".
 */
object BankFileParser {

    enum class Format { UNKNOWN, OFX, QIF, CSV, XML, HTML }

    /** True for the formats this client can actually read. XML/HTML are detected but unported — see the class note. */
    fun isSupported(format: Format) = format == Format.CSV || format == Format.OFX || format == Format.QIF

    /**
     * Guess the format from the content, then the file name. Content comes first for the ambiguous ones: some banks
     * name an HTML-table export ".xls" and an XML export ".xml"/".camt".
     */
    fun detect(fileName: String?, text: String): Format {
        val head = trimBom(text).trimStart()
        if (head.contains("<OFX", true) || head.contains("<STMTTRN", true)) return Format.OFX
        if (head.startsWith("<html", true) || head.contains("<table", true)) return Format.HTML
        if (head.contains("BkToCstmrStmt", true) || head.contains(":camt.", true) ||
            head.contains("<AccountMovement", true) ||
            (head.startsWith("<") && head.contains("<Ntry", true))
        ) return Format.XML

        val ext = (fileName ?: "").lowercase(Locale.ROOT)
        if (ext.endsWith(".ofx") || ext.endsWith(".qfx")) return Format.OFX
        if (ext.endsWith(".qif")) return Format.QIF
        if (ext.endsWith(".htm") || ext.endsWith(".html")) return Format.HTML
        if (ext.endsWith(".xml") || ext.endsWith(".camt")) return Format.XML
        if (ext.endsWith(".csv") || ext.endsWith(".tsv") || ext.endsWith(".txt")) return Format.CSV
        if (head.startsWith("!Type:", true)) return Format.QIF
        return if (text.contains(',') || text.contains(';') || text.contains('\t')) Format.CSV else Format.UNKNOWN
    }

    private fun trimBom(s: String) = if (s.isNotEmpty() && s[0] == '﻿') s.substring(1) else s

    // ---- OFX / QFX ----------------------------------------------------------------------------

    fun parseOfx(text: String): List<ImportedTransaction> {
        val list = mutableListOf<ImportedTransaction>()
        // Blocks are split on the opening tag; each piece then holds one transaction's fields.
        val blocks = text.split("<STMTTRN>", ignoreCase = true, limit = 0)
        for (i in 1 until blocks.size) {
            val block = blocks[i]
            val amountRaw = tag(block, "TRNAMT") ?: continue
            val dateRaw = tag(block, "DTPOSTED") ?: continue
            val amount = tryAmount(amountRaw) ?: continue
            val date = tryOfxDate(dateRaw) ?: continue
            val desc = tag(block, "NAME") ?: tag(block, "MEMO") ?: ""
            list.add(ImportedTransaction(date, amount, clean(desc)))
        }
        return list
    }

    /**
     * Read an OFX/SGML tag value: works whether the tag is closed (`<T>v</T>`) or open (`<T>v` up to the next tag
     * or newline), because real OFX files mix both.
     */
    private fun tag(block: String, tag: String): String? {
        val open = "<$tag>"
        var s = block.indexOf(open, ignoreCase = true)
        if (s < 0) return null
        s += open.length
        var end = s
        while (end < block.length && block[end] != '<' && block[end] != '\r' && block[end] != '\n') end++
        return block.substring(s, end).trim()
    }

    private fun tryOfxDate(raw: String): LocalDate? {
        val digits = raw.filter { it.isDigit() }
        if (digits.length < 8) return null
        return try {
            LocalDate.parse(digits.substring(0, 8), DateTimeFormatter.ofPattern("yyyyMMdd", Locale.ROOT))
        } catch (_: DateTimeParseException) {
            null
        }
    }

    // ---- QIF ----------------------------------------------------------------------------------

    fun parseQif(text: String): List<ImportedTransaction> {
        val list = mutableListOf<ImportedTransaction>()
        var date: LocalDate? = null
        var amount: Double? = null
        var payee = ""
        var memo = ""
        for (rawLine in text.split('\n')) {
            val line = rawLine.trimEnd('\r')
            if (line.isEmpty()) continue
            when (line[0]) {
                'D' -> tryLooseDate(line.substring(1).trim())?.let { date = it }
                'T', 'U' -> tryAmount(line.substring(1))?.let { amount = it }
                'P' -> payee = line.substring(1).trim()
                'M' -> memo = line.substring(1).trim()
                '^' -> {
                    val d = date
                    val a = amount
                    if (d != null && a != null) {
                        list.add(ImportedTransaction(d, a, clean(if (payee.isNotEmpty()) payee else memo)))
                    }
                    date = null; amount = null; payee = ""; memo = ""
                }
            }
        }
        return list
    }

    // ---- CSV ----------------------------------------------------------------------------------

    /**
     * Split CSV/TSV text into a header row + data rows, honouring quoted fields and auto-detecting the delimiter
     * (comma, semicolon or tab). Empty header when there are no rows.
     */
    fun readCsv(text: String): Pair<List<String>, List<List<String>>> {
        val lines = text.replace("\r\n", "\n").split('\n').filter { it.isNotBlank() }
        if (lines.isEmpty()) return emptyList<String>() to emptyList()
        val delimiter = detectDelimiter(lines[0])
        val all = lines.map { splitCsvLine(it, delimiter) }
        return all[0] to all.drop(1)
    }

    private fun detectDelimiter(header: String): Char {
        val tab = header.count { it == '\t' }
        val semi = header.count { it == ';' }
        val comma = header.count { it == ',' }
        return if (tab >= semi && tab >= comma && tab > 0) '\t' else if (semi > comma) ';' else ','
    }

    private fun splitCsvLine(line: String, delimiter: Char): List<String> {
        val fields = mutableListOf<String>()
        val sb = StringBuilder()
        var inQuotes = false
        var i = 0
        while (i < line.length) {
            val ch = line[i]
            when {
                inQuotes -> when {
                    ch == '"' && i + 1 < line.length && line[i + 1] == '"' -> { sb.append('"'); i++ }
                    ch == '"' -> inQuotes = false
                    else -> sb.append(ch)
                }
                ch == '"' -> inQuotes = true
                ch == delimiter -> { fields.add(sb.toString().trim()); sb.clear() }
                else -> sb.append(ch)
            }
            i++
        }
        fields.add(sb.toString().trim())
        return fields
    }

    /**
     * Turn CSV data rows into transactions using the caller-chosen column indices. The amount can come from one
     * signed column, or from separate debit/credit columns (debit = out, credit = in).
     *
     * ⚠️ A row that does not parse is SKIPPED, not thrown on, so one bad line never fails a whole import — banks
     * put subtotal and balance-carried-forward rows in the middle of real ones.
     */
    fun parseCsv(
        rows: List<List<String>>,
        dateCol: Int,
        descCol: Int,
        amountCol: Int,
        debitCol: Int? = null,
        creditCol: Int? = null,
    ): List<ImportedTransaction> {
        val list = mutableListOf<ImportedTransaction>()
        for (r in rows) {
            fun col(i: Int?): String? = if (i != null && i >= 0 && i < r.size) r[i] else null
            val date = tryLooseDate((col(dateCol) ?: "").trim()) ?: continue

            val amount: Double
            if (debitCol != null || creditCol != null) {
                val debit = tryAmount(col(debitCol) ?: "")?.let { kotlin.math.abs(it) } ?: 0.0
                val credit = tryAmount(col(creditCol) ?: "")?.let { kotlin.math.abs(it) } ?: 0.0
                amount = credit - debit
            } else {
                amount = tryAmount(col(amountCol) ?: "") ?: continue
            }
            if (amount == 0.0) continue

            list.add(ImportedTransaction(date, amount, clean(col(descCol) ?: "")))
        }
        return list
    }

    // ---- shared helpers -----------------------------------------------------------------------

    /**
     * Parse a money string tolerant of currency symbols, spaces, thousands separators, both decimal conventions
     * (1,234.56 and 1.234,56) and parenthesised negatives. Null when it is not a number at all.
     */
    fun tryAmount(raw: String): Double? {
        if (raw.isBlank()) return null
        val s = raw.trim()
        val negative = (s.startsWith("(") && s.endsWith(")")) || s.contains('-')
        var cleaned = s.filter { it.isDigit() || it == '.' || it == ',' }
        if (cleaned.isEmpty()) return null

        val lastDot = cleaned.lastIndexOf('.')
        val lastComma = cleaned.lastIndexOf(',')
        if (lastDot >= 0 && lastComma >= 0) {
            // Whichever separator comes LAST is the decimal point; the other is a thousands separator. This is what
            // makes "1.234,56" and "1,234.56" both read as the same amount instead of one of them as 1.23.
            val decimalSep = if (lastDot > lastComma) '.' else ','
            val thousandsSep = if (decimalSep == '.') ',' else '.'
            cleaned = cleaned.replace(thousandsSep.toString(), "").replace(decimalSep, '.')
        } else if (lastComma >= 0) {
            // Only commas: a decimal if it looks like one ("12,50"), otherwise a thousands separator ("1,234").
            cleaned = if (cleaned.length - lastComma - 1 <= 2 && cleaned.count { it == ',' } == 1) {
                cleaned.replace(',', '.')
            } else {
                cleaned.replace(",", "")
            }
        }
        val parsed = cleaned.toBigDecimalOrNull() ?: return null
        val abs = parsed.abs()
        return if (negative) abs.negate().toDouble() else abs.toDouble()
    }

    private fun String.toBigDecimalOrNull(): BigDecimal? = try {
        BigDecimal(this)
    } catch (_: NumberFormatException) {
        null
    }

    // ⚠️ ORDER IS LOAD-BEARING. ISO and dd/MM come before MM/dd so European statements read correctly; 03/04 is a
    // real date under both, so whichever is tried first silently wins. Same list, same order, as the C# parser.
    private val dateFormats = listOf(
        "yyyy-MM-dd", "yyyy/MM/dd", "dd/MM/yyyy", "dd.MM.yyyy", "dd-MM-yyyy",
        "MM/dd/yyyy", "d/M/yyyy", "M/d/yyyy", "yyyyMMdd", "dd/MM/yy", "MM/dd/yy",
    )

    /**
     * Parse a date across the common bank formats, dropping a trailing time component (Revolut's
     * "2026-07-01 12:09:37"). Genuinely ambiguous values are shown to the user in the review step rather than
     * guessed at more cleverly here.
     */
    fun tryLooseDate(raw: String): LocalDate? {
        if (raw.isBlank()) return null
        val s = raw.trim().replace("'", "/")
        val sep = s.indexOfFirst { it == ' ' || it == 'T' || it == 't' }
        val datePart = if (sep > 0) s.substring(0, sep) else s
        for (f in dateFormats) {
            try {
                return LocalDate.parse(datePart, DateTimeFormatter.ofPattern(f, Locale.ROOT))
            } catch (_: DateTimeParseException) {
                // try the next one
            }
        }
        return try {
            LocalDate.parse(datePart)
        } catch (_: DateTimeParseException) {
            null
        }
    }

    /** Collapse runs of whitespace — bank descriptions are full of padding. */
    private fun clean(s: String) = s.split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ").trim()

    // ---- column guessing ----------------------------------------------------------------------

    /**
     * A first guess at which column is which, so the mapper opens already filled in for the common exports. Only a
     * guess: every one is user-changeable, because the cost of a wrong guess nobody can correct is a mis-imported
     * statement.
     */
    fun guessColumns(headers: List<String>): ColumnGuess {
        fun find(vararg words: String): Int = headers.indexOfFirst { h ->
            val t = h.lowercase(Locale.ROOT)
            words.any { t.contains(it) }
        }
        val debit = find("debit", "withdrawal", "money out", "paid out", "outflow")
        val credit = find("credit", "deposit", "money in", "paid in", "inflow")
        return ColumnGuess(
            date = find("date", "дата", "posted", "booking").coerceAtLeast(0),
            description = find("description", "narrative", "details", "payee", "reason", "merchant", "name")
                .let { if (it >= 0) it else 1.coerceAtMost(headers.lastIndex.coerceAtLeast(0)) },
            amount = find("amount", "sum", "value", "сума"),
            debit = debit.takeIf { it >= 0 },
            credit = credit.takeIf { it >= 0 },
        )
    }

    /** Which column holds what. When [debit]/[credit] are set they win over [amount] — see [parseCsv]. */
    data class ColumnGuess(
        val date: Int,
        val description: Int,
        val amount: Int,
        val debit: Int?,
        val credit: Int?,
    )
}
